pkg/fakedri/fakedri.go

```diff
@@ -0,0 +1,529 @@
+		sysfsPath, devfsPath)
+
+	opts.dirs, opts.files, opts.devs, opts.symls = 0, 0, 0, 0
+	for i := 0; i < opts.DevCount; i++ {
+		if err := addSysfsBusTree(sysfsPath, &opts, i); err != nil {
+			klog.Fatalf("Dev-%d sysfs bus tree generation failed: %v", i, err)
+		}
+
+		if err := addSysfsDriTree(sysfsPath, &opts, i); err != nil {
+			klog.Fatalf("Dev-%d sysfs tree generation failed: %v", i, err)
+		}
+
+		if err := addDevfsDriTree(devfsPath, &opts, i); err != nil {
+			klog.Fatalf("Dev-%d devfs tree generation failed: %v", i, err)
+		}
+
+		if err := addDebugfsDriTree(sysfsPath, &opts, i); err != nil {
+			klog.Fatalf("Dev-%d debugfs tree generation failed: %v", i, err)
+		}
+	}
+
+	klog.V(1).Infof("Done, created %d dirs, %d devices, %d files and %d symlinks.", opts.dirs, opts.devs, opts.files, opts.symls)
+
+	makeXelinkSideCar(opts)
+}
+
+func makeXelinkSideCar(opts GenOptions) {
+	topology := opts.Capabilities["connection-topology"]
+	gpus := opts.DevCount
+	tiles := opts.TilesPerDev
+	connections := opts.Capabilities["connections"]
+
+	if topology == fullyConnected {
+		saveSideCarFile(buildConnectionList(gpus, tiles))
+	} else if connections != "" {
+		saveSideCarFile(connections)
+	} else {
+		return
+	}
+
+	klog.V(1).Infof("XELINK: generated xelink sidecar label file, using (GPUs: %d, Tiles: %d, Topology: %s)", gpus, tiles, topology)
+}
+
+func buildConnectionList(gpus, tiles int) string {
+	var nodes = make([]string, 0)
+
+	for mm := 0; mm < gpus; mm++ {
+		for nn := 0; nn < tiles; nn++ {
+			nodes = append(nodes, fmt.Sprintf("%d.%d", mm, nn))
+		}
+	}
+
+	var links = make(map[string]bool, 0)
+
+	var smap = make([]string, 0)
+
+	for _, from := range nodes {
+		for _, to := range nodes {
+			if to == from {
+				continue
+			}
+
+			link := fmt.Sprintf("%s-%s", to, from)
+
+			reverselink := fmt.Sprintf("%s-%s", from, to)
+			if _, exists := links[reverselink]; !exists {
+				links[link] = true
+
+				smap = append(smap, link)
+			}
+		}
+	}
+
+	return strings.Join(smap, "_")
+}
+
+func saveSideCarFile(connections string) {
+	// Get user-specific temp directory
+	filePath := filepath.Join("/etc/kubernetes/node-feature-discovery/features.d", "xpum-sidecar-labels.txt")
+
+	// Safely create file in the temp directory
+	f, err := os.Create(filePath)
+	if err != nil {
+		klog.Fatalf("Failed to create file: %v", err)
+	}
+	defer f.Close()
+
+	line := fmt.Sprintf("xpumanager.intel.com/xe-links=%s", connections[:min(len(connections), maxK8sLabelSize)])
+	klog.V(1).Info(line)
+
+	if _, err := f.WriteString(line + "\n"); err != nil {
+		panic(err)
+	}
+
+	index := 2
+
+	for i := maxK8sLabelSize; i < len(connections); i += (maxK8sLabelSize - 1) {
+		line := fmt.Sprintf("xpumanager.intel.com/xe-links%d=Z%s", index, connections[i:min(len(connections), i+maxK8sLabelSize-1)])
+		klog.V(1).Info(line)
+
+		if _, err := f.WriteString(line + "\n"); err != nil {
+			panic(err)
+		}
+
+		index++
+	}
+}
+
+func MakeOptions(opts GenOptions) GenOptions {
```

Comment:
This does not create or even change the options, and seems internal:
```suggestion
func verifyOptions(opts GenOptions) GenOptions {
```

Reply:
changed to verifyOptions